core/server/master/src/main/java/alluxio/master/file/meta/UfsSyncPathCache.java

```diff
@@ -33,18 +36,30 @@
 public final class UfsSyncPathCache {
   private static final Logger LOG = LoggerFactory.getLogger(UfsSyncPathCache.class);
 
-  /** Number of paths to cache. */
-  private static final int MAX_PATHS =
-      Configuration.getInt(PropertyKey.MASTER_UFS_PATH_CACHE_CAPACITY);
-
   /** Cache of paths which have been synced. */
-  private final Cache<String, SyncTime> mCache;
+  public final Cache<String, SyncTime> mCache;
 
   /**
    * Creates a new instance of {@link UfsSyncPathCache}.
+   *
+   * TODO(jiacheng): One potential issue is those NOT_SYNCED entries are not evicted
+   *   So the memory will not be released if BOTH:
+   *   (1) Many forced sync entries; (2) No sync really happens
+   *   However that should be a really rare case?
```

Comment:
See the trade-off here, I figure this is acceptable

Reply:
Whenever we put the non-evictable `FORCED_SYNC` into the Cache, here's what will happen:
1. If the user retries with GetStatus on the file, a sync will be triggered and the FORCED_SYNC entry will be replaced with an evictable normal entry.
2. If the user doesn't retry but does something like `ListStatus` on the parent dir, e.g. using `alluxio ls -f` command, this file will also be accessed and a sync should happen on the file. I need to double check to make sure the FORCED_SYNC entry is removed.
3. If the user `ListStatus` on the parent dir but the file is already removed from the UFS, I need to double check and make sure the `FORCED_SYNC` entry is removed.
4. If the user doesn't do anything at all on this path, this `FORCED_SYNC` entry will stay there forever and the cache will essentially contain more entries than it should.

Reply:
I guess my only concern would be:
- Do certain clients use the results from listStatus to read the child files (i.e. they don't need to call the child)?
  - The reasoning behind this would be, that adding the invalidation to a path, does not mean the parent will necessary execute a metadata sync on the next list status call.
  - So then if the reader is using the result of the listStatus it may still see the outdated data until the sync time runs out on the parent.

Reply:
This is a valid concern. I tried to patch this but all the methods I thought of introduce new problems. 

Current behavior first:
After we call `forceNextSync(filePath)`, the next `ls/getStatus` on the file will trigger a sync. But a `ls` on the parent dir doesn't know this so there are problems.

If I somehow mark the parent dir to sync, then the question becomes what do I do if the user lists the parent of parent. One mitigation to that would be to just mark the `directChildrenLoaded` flag on the directory and when I recursively list a directory, make sure i always check this flag on all children dirs so I never miss a hint propagated from a file.

One step further, do we need durability on this operation? It makes my head explode to think about restarts or master failovers, given all the complexity introduced by the parent directories already.

Given all the problems here, what about taking a step back and forcing a sync on the file immediately instead of on the next access? Sync on the parent directory sounds more expensive than sync-ing on that file already. We originally wanted to save that one file sync, now that turns out to cost us more.

Then we turn a blind eye to the gap between this sync and the next time this file is accessed. Because if the file is modified again, it's the responsibility of the next access to update the file metadata yet again.

Reply:
It is reasonable. But noting the fact that from the client side, no matter whether we sync immediately or not, the client always needs to discard the bytes that it has read, and creates a new FileInStream. Then it has to retry reading the same file next time.

Reply:
Look good to me.

Reply:
I'll leave this comment here unresolved so others will know why we made this decision if they trace to this PR.